File: src/action/Sleep.php
<?php


namespace Ething\Action;
	
class Sleep extends Action {
	
	
	// attributes validation function
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 'duration' => null ),
			$attributes
		);
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'duration':
				
					if(!is_int($attributes[$key]) || $attributes[$key]<1 || $attributes[$key]>3600)
						throw new \Exception("field '{$key}' must be an intger greater than 1 and lower than .");
					
					break;
				
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		return true; 
	}
	
	
	protected function call(\Ething\Event\Signal $signal){
		\sleep($this->duration);
	}
	
}

File: ething/rule/action/Sleep.py
# coding: utf-8
from .. import Action
import time


class Sleep(Action):
    
    @staticmethod
    def validate (attributes, context):
        
        attributes.setdefault('duration', None)
        
        for key in attributes:
            
            
            if key == 'duration':
                
                duration = attributes[key]
                
                if not((isinstance(duration, int) or isinstance(duration, float)) and duration > 0 and duration < 3600 ):
                    raise Exception("duration must be a number of seconds greater than 0 and lower than 1 hour.");
                
            else:
                raise Exception("%s: invalid" % key)
        
        return True
    
    
    
    def call(self, signal):
        time.sleep(self['duration'])
    
